File: 2025/day-06/part2.py
import sys
from math import prod

if len(sys.argv) > 1:
	INPUT = sys.argv[1]
else:
	INPUT = "input2.txt"
errored = []
with open(INPUT) as f:
	# stupid ass shit WITH spaces
	lines = [line.strip("\n") for line in f]
	max_len = max(len(line) for line in lines)

	# padding to ensure safe access
	rows = [line.ljust(max_len) for line in lines]

	ops = list(rows.pop())

	total = 0
	newdigits = []
	alldigits = []
	for col in reversed(range(len(ops))):
		digits = ""
		for row in rows:
			digit = row[col]
			if digit == " ":
				digits += ""
			else:
				digits += digit
		if digits == "":
			newdigits = []
		else:
			newdigits.append(int(digits))
		op = ops[col]
		if op == "*":
			total += prod(newdigits)
			alldigits.append(newdigits)
		elif op == "+":
			total += sum(newdigits)
			alldigits.append(newdigits)
		elif op == " ":
			pass

if not (
	[356, 24, 1] in alldigits
	and [4, 431, 623] in alldigits
	and [175, 581, 32] in alldigits
	and [8, 248, 369] in alldigits
):
	print(f"error in {alldigits}")
# expected: 1058 + 3253600 + 625 + 8544 = 3263827
print(total)

File: 2025/day-06/src/part2.rs
use itertools::{Itertools, max};
use pad::PadStr;

#[tracing::instrument]
pub fn process(_input: &str) -> miette::Result<String> {
	let mut lines = _input
		.lines()
		.map(|line| line.trim_matches('\n'))
		.collect_vec();
	let ops = lines
		.pop()
		.expect("operators trailing")
		.chars()
		.collect_vec();
	let max_len = max(lines.iter().map(|line| line.len())).expect("maximum length");
	let rows = lines
		.iter_mut()
		.map(|line| line.pad_to_width_with_alignment(max_len, pad::Alignment::Left))
		.collect_vec();
	let mut cur_digits: Vec<usize> = Vec::with_capacity(lines.len());
	let total: usize = (0..ops.len())
		.rev()
		.filter_map(|col| {
			let digits: String = rows
				.iter()
				.filter_map(|row| row.chars().nth(col).filter(|&digit| digit != ' '))
				.collect();
			if digits.is_empty() {
				cur_digits.clear();
			} else {
				cur_digits.push(digits.parse::<usize>().unwrap());
			}
			match ops[col] {
				'+' => Some(cur_digits.iter().sum::<usize>()),
				'*' => Some(cur_digits.iter().product::<usize>()),
				_ => None,
			}
		})
		.sum();
	Ok(total.to_string())
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_process() -> miette::Result<()> {
		let input = "123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
";
		assert_eq!("3263827", process(input)?);
		Ok(())
	}
}
